### DISCUSSION TITLE
How to infer models trained on custom datasets

### DISCUSSION THREAD

**Discussion by Shin-ichi-Takayama** [speechbrain/speechbrain#1931] [Q&A] [answered]
Hi everyone, I have a question.
I have created an original model with a custom dataset.
However, I don't know how to infer using the model I created.

### What I have done
I created a custom dataset. Then I created the model using `sepformer-customdataset.yaml`.
I did the following to infer with the model I created
- I copied the directory `results/sepformer-custom/1000/save/CKPT+2023-03-30+12-04-55+00` to the same directory as `speech_separation.py` below.
- Copy `sepformer-customdataset.yaml` to the `CKPT+2023-03-30+12-04-55+00` directory and rename it to `hyperparams.yaml`.
- Run the following code

speech_separation.py
```python
from speechbrain.pretrained import SepformerSeparation as separator
import torchaudio

#model = separator.from_hparams(source="speechbrain/sepformer-wsj02mix", savedir='pretrained_models/sepformer-wsj02mix')
model = separator.from_hparams(source="speechbrain/sepformer-wsj02mix", savedir='CKPT+2023-03-30+12-04-55+00')

# for custom file, change path
est_sources = model.separate_file(path='speechbrain/sepformer-wsj02mix/test_mixture.wav') 

torchaudio.save("source1hat.wav", est_sources[:, :, 0].detach().cpu(), 8000)
torchaudio.save("source2hat.wav", est_sources[:, :, 1].detach().cpu(), 8000)
```

### Error that occurred
```
Traceback (most recent call last):
  File "/root/speech_separate.py", line 5, in <module>
    model = separator.from_hparams(source="speechbrain/sepformer-wsj02mix", savedir='CKPT+2023-03-30+12-04-55+00')
  File "/usr/local/lib/python3.9/dist-packages/speechbrain/pretrained/interfaces.py", line 399, in from_hparams
    hparams = load_hyperpyyaml(fin, overrides)
  File "/usr/local/lib/python3.9/dist-packages/hyperpyyaml/core.py", line 157, in load_hyperpyyaml
    yaml_stream = resolve_references(yaml_stream, overrides, overrides_must_match)
  File "/usr/local/lib/python3.9/dist-packages/hyperpyyaml/core.py", line 325, in resolve_references
    _walk_tree_and_resolve("root", preview, preview, file_path)
  File "/usr/local/lib/python3.9/dist-packages/hyperpyyaml/core.py", line 372, in _walk_tree_and_resolve
    current_node[k] = _walk_tree_and_resolve(sub_key, sub_node, tree, file_path)
  File "/usr/local/lib/python3.9/dist-packages/hyperpyyaml/core.py", line 380, in _walk_tree_and_resolve
    raise ValueError(f"'{key}' is a !PLACEHOLDER and must be replaced.")
ValueError: 'data_folder' is a !PLACEHOLDER and must be replaced.
```
[hyperparams.zip](https://github.com/speechbrain/speechbrain/files/11196891/hyperparams.zip)

I attached the hyperparams.yaml as zip file.
Please let me know how to respond.
Best regards.
Translated with www.DeepL.com/Translator (free version)

**Comment by asumagic** [comment, marked as answer]
Maybe you can use the _inference_ [`hyperparameters.yaml`](https://huggingface.co/speechbrain/sepformer-wsj02mix/blob/main/hyperparams.yaml) from the `speechbrain/sepformer-wsj02mix` HuggingFace model directly, adjusting for changes you may have done.

Alternatively, you can edit yours to make it suitable for inference. For one, you should simply be able to remove that `data_folder` line for inference as it is only relevant for training.

**Reply by Shin-ichi-Takayama** [reply]
@AsuMagic 
Thank you for your response.
I have removed the unwanted and unnecessary items that are not needed for inference in the yaml file.
However, I got a new error.
```
Traceback (most recent call last):
  File "/root/speech_separate.py", line 5, in <module>
    model = separator.from_hparams(source="speechbrain/sepformer-wsj02mix", savedir='CKPT+2023-03-30+12-04-55+00')
  File "/usr/local/lib/python3.9/dist-packages/speechbrain/pretrained/interfaces.py", line 402, in from_hparams
    pretrainer = hparams["pretrainer"]
KeyError: 'pretrainer'
```

Best regards.

**Reply by asumagic** [reply]
The `hyperparameters.yaml` I linked seems to cover that too, try adding:

```yaml


pretrainer: !new:speechbrain.utils.parameter_transfer.Pretrainer
  loadables:
    masknet: !ref <MaskNet>
    encoder: !ref <Encoder>
    decoder: !ref <Decoder>
```

If you get further errors, try comparing against that inference yaml I linked as it is meant to be used with the same interface and type of model as you are using.

**Reply by Shin-ichi-Takayama** [reply]
@AsuMagic 
Thank you for your reply.
My apologies. I had overlooked that part.
I was able to get it to work without any errors.
However, there is one problem.
When I input audio with a sampling frequency of 16 kHz, it seems to output 8 kHz audio.
Is there any way to output at 16 kHz?

Best regards.

**Reply by asumagic** [reply]
Ah, it seems that `customdataset` _is_ trained for 16kHz but that the HuggingFace model is trained for 8kHz, so the hyperparameters file would actually be incorrect in your case. In the file, try editing `sample_rate` to `16000`. I didn't notice any other difference so hopefully that should suffice.

**Reply by Shin-ichi-Takayama** [reply]
@AsuMagic 
Thank you for your response.
I had set the sampling frequency to 8 kHz when I saved the data.
So I rewrote the code as follows.

```python
from speechbrain.pretrained import SepformerSeparation as separator
import torchaudio

#model = separator.from_hparams(source="speechbrain/sepformer-wsj02mix", savedir='pretrained_models/sepformer-wsj02mix')
model = separator.from_hparams(source="speechbrain/sepformer-wsj02mix", savedir='CKPT+2023-03-30+12-04-55+00')

# for custom file, change path
#est_sources = model.separate_file(path='speechbrain/sepformer-wsj02mix/test_mixture.wav') 
est_sources = model.separate_file(path='test_mixture_16kHz.wav') 

torchaudio.save("source1hat.wav", est_sources[:, :, 0].detach().cpu(), 16000)
torchaudio.save("source2hat.wav", est_sources[:, :, 1].detach().cpu(), 16000)
```

Thanks to you.
